src/lib.rs

```diff
@@ -65,6 +73,402 @@ mod macros {
+        {
+            use zeroize::Zeroize;
+            self.zeroize();
+        }
+    }
+}
+
+/// The ciphertext computed by the encapsulator.
+#[derive(Debug)]
+pub struct Ciphertext([u8; CRYPTO_CIPHERTEXTBYTES]);
+
+impl Ciphertext {
+    pub fn as_array(&self) -> &[u8; CRYPTO_CIPHERTEXTBYTES] {
+        &self.0
+    }
+}
+
+impl AsRef<[u8]> for Ciphertext {
+    fn as_ref(&self) -> &[u8] {
+        self.0.as_ref()
+    }
+}
+
+/// The shared secret computed by the KEM. Returned from both the
+/// encapsulator and decapsulator.
+#[derive(Debug, Eq, PartialEq)]
+pub struct SharedSecret<'a>(KeyBuffer<'a, CRYPTO_BYTES>);
+
+impl<'a> SharedSecret<'a> {
+    pub fn as_array(&self) -> &[u8; CRYPTO_BYTES] {
+        self.0.as_ref()
+    }
+}
+
+impl<'a> AsRef<[u8]> for SharedSecret<'a> {
+    fn as_ref(&self) -> &[u8] {
+        self.0.as_ref()
+    }
+}
+
+#[cfg(feature = "zeroize")]
+impl<'a> zeroize::Zeroize for SharedSecret<'a> {
+    fn zeroize(&mut self) {
+        self.0.zeroize();
+    }
+}
+
+#[cfg(feature = "zeroize")]
+impl<'a> zeroize::ZeroizeOnDrop for SharedSecret<'a> {}
+
+impl<'a> Drop for SharedSecret<'a> {
+    fn drop(&mut self) {
+        #[cfg(feature = "zeroize")]
+        {
+            use zeroize::Zeroize;
+            self.zeroize();
+        }
+    }
+}
+
+/// KEM Keypair generation.
+///
+/// Generate a public and secret key.
+/// The public key is meant to be shared with any party,
+/// but access to the secret key must be limited to the generating party.
+pub fn keypair<
+    'public,
+    'secret,
+    P: Into<KeyBuffer<'public, CRYPTO_PUBLICKEYBYTES>> + 'public,
+    S: Into<KeyBuffer<'secret, CRYPTO_SECRETKEYBYTES>> + 'secret,
+    R: CryptoRng + RngCore,
+>(
+    public_key_buf: P,
+    secret_key_buf: S,
+    rng: &mut R,
+) -> (PublicKey<'public>, SecretKey<'secret>) {
+    let mut public_key_buf = public_key_buf.into();
+    let mut secret_key_buf = secret_key_buf.into();
+
+    operations::crypto_kem_keypair(public_key_buf.as_mut(), secret_key_buf.as_mut(), rng);
+
+    (PublicKey(public_key_buf), SecretKey(secret_key_buf))
+}
+
+#[cfg(feature = "alloc")]
+pub fn keypair_boxed<R: CryptoRng + RngCore>(
+    rng: &mut R,
+) -> (PublicKey<'static>, SecretKey<'static>) {
+    let mut pk_buffer = Box::new([0u8; CRYPTO_PUBLICKEYBYTES]);
+    let mut sk_buffer = Box::new([0u8; CRYPTO_SECRETKEYBYTES]);
+
+    operations::crypto_kem_keypair(&mut pk_buffer, &mut sk_buffer, rng);
+
+    (PublicKey(pk_buffer.into()), SecretKey(sk_buffer.into()))
+}
+
+/// KEM Encapsulation.
+///
+/// Given a public key `public_key`, compute a shared key.
+/// The returned ciphertext should be sent back to the entity holding
+/// the secret key corresponding to public key given here, so they can compute
+/// the same shared key.
+pub fn encapsulate<
+    'shared_secret,
+    S: Into<KeyBuffer<'shared_secret, CRYPTO_BYTES>>,
+    R: CryptoRng + RngCore,
+>(
+    public_key: &PublicKey<'_>,
+    shared_secret_buf: S,
```

Comment:
I find these two `shared_secret_buf` for the encapsulate & decapsulate pretty un-ergonomic.
Maybe a variant without them that just returns them as owned arrays?
This probably again does not work well with `ZeroizeOnDrop` though...

Reply:
If we don't provide the storage, then owned arrays or heap allocation are the only ways to go IMO. So yeah, either than convenience version is only available on `alloc` or we sidestep the `zeroize` feature. Given that the array version would require more types/fundamental changes to the code I already have here, I'm leaning towards separate `encapsulate_boxed` versions available with `alloc` only.